File: jm/diag/kwp1281.cpp
#include "kwp1281.h"
#include "../utils.h"

namespace JM
{
namespace Diag
{

KWP1281::KWP1281()
: _frameCounter(0)
{

}

KWP1281::~KWP1281()
{

}

boost::uint8_t KWP1281::frame_counter_increment()
{
    return ++(_frameCounter);
}

std::size_t KWP1281::pack(const boost::uint8_t *src,
                          std::size_t srcLength,
                          boost::uint8_t *tar,
                          std::size_t tarLength,
                          boost::system::error_code &ec)
{
    ec = boost::system::error_code();
    tar[0] = low_byte(srcLength + 20);
    tar[1] = frame_counter_increment();

    memcpy(tar + 2, src, srcLength);
    tar[srcLength + 2] = FRAME_END;
    return srcLength + 3;
}

std::size_t KWP1281::unpack(const boost::uint8_t *src,
                            std::size_t srcLength,
                            boost::uint8_t *tar,
                            std::size_t tarLength,
                            boost::system::error_code &ec)
{
    ec = boost::system::error_code();
    memcpy(tar, src + 1, srcLength - 2);
    return srcLength - 2;
}

boost::system::error_code
KWP1281::config(const Options &opts, boost::system::error_code &ec)
{
    memcpy(&_opts, &opts, sizeof (Options));
    ec = boost::system::error_code();
    return ec;
}

}
}

File: JM/Diag/KWP1281.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using JM.Core;

namespace JM.Diag
{
    class KWP1281 : IProtocol
    {
        private byte frameCounter;
        protected KWP1281Options options;
        public static byte FRAME_END = 0x03;
        KWP1281()
        {
            frameCounter = 0;
        }

        private byte frameCounterIncrement()
        {
            return ++frameCounter;
        }

        public byte[] Pack(byte[] data, int offset, int count)
        {
            byte[] result = new byte[count + 3];
            result[0] = Utils.LowByte(count + 2);
            result[1] = frameCounterIncrement();

            Array.Copy(data, offset, result, 2, count);
            result[result.Length - 1] = FRAME_END;
            return result;
        }

        public byte[] Unpack(byte[] data, int offset, int count)
        {
            byte[] result = new byte[count - 2];
            Array.Copy(data, offset + 1, result, 0, count - 2);
            return result;
        }

        public void Config(KWP1281Options opts)
        {
            options = opts;
        }
    }
}
